README.md

```diff
@@ -4,6 +4,301 @@
+*   per\_device\_batch\_size: decoding batch size per device (1 TPU chip = 1 device)
+*   max\_prefill\_predict\_length: Maximum length for the prefill when doing autoregression
+*   max\_target\_length: Maximum sequence length
+*   model\_name: Model name
+*   ici\_fsdp\_parallelism: The number of shards for FSDP parallelism
+*   ici\_autoregressive\_parallelism: The number of shards for autoregressive parallelism
+*   ici\_tensor\_parallelism: The number of shards for tensor parallelism
+*   weight\_dtype: Weight data type (e.g. bfloat16)
+*   scan\_layers: Scan layers boolean flag
+
+Note: these flags are from [MaxText config](https://github.com/google/maxtext/blob/f9e04cdc1eec74a0e648411857c09403c3358461/MaxText/configs/base.yml)
+
+
+## Step 5: Send test request to JetStream MaxText server
+
+```bash
+cd ~
+python JetStream/jetstream/tools/requester.py
+```
+
+The output will be similar to the following:
+
+```bash
+Sending request to: dns:///[::1]:9000
+Prompt: Today is a good day
+Response:  to be a fan
+```
+
+## Step 6: Run benchmarks with JetStream MaxText server
+
+Note: The JetStream MaxText Server is not running with quantization optimization in Step 3. To get best benchmark results, we need to enable quantization (Please use AQT trained or fine tuned checkpoints to ensure accuracy) for both weights and KV cache, please add the quantization flags and restart the server as following:
+
+```bash
+# Enable int8 quantization for both weights and KV cache
+export QUANTIZATION=int8
+export QUANTIZE_KVCACHE=true
+
+# For Gemma 7b model, change per_device_batch_size to 12 to optimize performance. 
+export PER_DEVICE_BATCH_SIZE=12
+
+cd ~/maxtext
+python MaxText/maxengine_server.py \
+MaxText/configs/base.yml \
+tokenizer_path=${TOKENIZER_PATH} \
+load_parameters_path=${LOAD_PARAMETERS_PATH} \
+max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
+max_target_length=${MAX_TARGET_LENGTH} \
+model_name=${MODEL_NAME} \
+ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
+ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
+ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
+scan_layers=${SCAN_LAYERS} \
+weight_dtype=${WEIGHT_DTYPE} \
+per_device_batch_size=${PER_DEVICE_BATCH_SIZE} \
+quantization=${QUANTIZATION} \
+quantize_kvcache=${QUANTIZE_KVCACHE}
+```
+
+### Benchmarking Gemma-7b
+
+```bash
+# For Gemma 7b, plz use Gemma tokenizer (must use a tokenizer that matches your model; double check if you are using tokenizer.gemma) and ShareGPT dataset; add `--warmup-first` flag for your 1st run to warmup the server.
+# Activate the python virtual environment we created in Step 2.
+cd ~
+source .env/bin/activate
+
+# download dataset
+wget https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json
+
+# run benchmark with the downloaded dataset and the tokenizer in maxtext
+# You can control the qps by setting `--request-rate`, the default value is inf.
```

Comment:
What is the effect of `--request-rate` being set to inf?

Reply:
dispatch all requests at once.